File: src/lib/fonts.ts
import raw from './fonts.generated.json'

type Metrics = {
  id: string
  family: string
  role: 'display' | 'text'
  file: string
  variable: boolean
  wght: number[]
  bytes: number
  upm: number
  capR: number
  ascR: number
  descR: number
  adv: Record<string, number>
}

export const METRICS = raw as unknown as Record<string, Metrics>

/**
 * Насколько вордмарк уезжает под нижнюю кромку вьюпорта.
 *
 * Слово обрезается ровно на CROP высоты литеры. Чтобы попасть точно,
 * считаем от реальных метрик гарнитуры, а не подбираем число на глаз:
 *   базовая линия от низа строки (line-height: 1) = 0.5 − (asc + desc) / 2
 *   плюс отрезаемая часть литеры                  = CROP × capHeight
 */
const CROP = 0.34

export const wordmarkDrop = (m: Metrics) =>
  +(0.5 - (m.ascR + m.descR) / 2 + CROP * m.capR).toFixed(4)

/** Сумма ширин Т-Е-Р-М-А в эмах — по ней вписываем кегль в ширину экрана. */
export const wordmarkSum = (m: Metrics) =>
  +Object.values(m.adv).reduce((a, b) => a + b, 0).toFixed(4)

export type Pair = {
  id: string
  title: string
  display: string
  /** Вес, которым набирается вордмарк и крупные заголовки. */
  displayWeight: number
  text: string
  foundry: string
  designer: string
  link: string
  price: string
  licence: string
  cyr: string
  rationale: string
}

/**
 * Кандидаты на замену. Forum, Prata и Tenor Sans выведены из проекта —
 * арт-директор забраковал их как самые заезженные «премиальные»
 * гарнитуры русского веба.
 *
 * Все три пары ниже отрисовываются настоящими файлами: витрина должна
 * показывать кадр, а не описание. Платные кандидаты из русских студий
 * перечислены в CLAUDE.md — их файлы за периметром сборки недоступны.
 */
export const PAIRS: Pair[] = [
  {
    id: 'golos',
    title: 'Golos Text 900 × Golos Text 400',
    display: 'golos',
    displayWeight: 900,
    text: 'golos',
    foundry: 'ParaType',
    designer: 'Александра Королькова, Виталий Кузьмин',
    link: 'https://www.paratype.ru/fonts/pt/golos-text',
    price: 'Бесплатно, SIL OFL 1.1 — включая коммерческое использование',
    licence: 'SIL Open Font License 1.1',
    cyr: 'Кириллица первична: гарнитура рисовалась под русский текст, латиница добавлена к ней, а не наоборот. По cmap 68 из 68 знаков, включая ₽ и №.',
    rationale:
      'Это то, что стоит в бою прямо сейчас — по вашему указанию, до выбора новой пары. ' +
      'Одна гарнитура в двух крайних весах вместо пары: 900 на вордмарк и крупные ' +
      'заголовки, 400 на текст. Расчёт на то, что контраст даёт вес и кегль, а не ' +
      'смена рисунка. У Golos прямые вертикали, крупный очковый размер и очень ровный ' +
      'ритм — на кегле 350 px тяжёлое начертание держится как вырубка по камню, без ' +
      'ложной парадности, которой грешат антиквы с Тильды. Минус честный: характера ' +
      'у неё меньше, чем у выделенной акциденции, — это рабочая нейтральность, а не ' +
      'высказывание.',
  },
  {
    id: 'unbounded',
    title: 'Unbounded × Onest',
    display: 'unbounded',
    displayWeight: 500,
    text: 'onest',
    foundry: 'NaN × Onest',
    designer: 'Luke Prowse и др. · Дмитрий Волошин, Андрей Кудрявцев',
    link: 'https://fonts.google.com/specimen/Unbounded',
    price: 'Бесплатно, SIL OFL 1.1',
    licence: 'SIL Open Font License 1.1',
    cyr: 'Кириллица входит в семейство с релиза и нарисована вместе с латиницей, но студия западная — это не кириллица-первоисточник. По cmap 68 из 68.',
    rationale:
      'Геометрический гротеск с очень широкими прописными и почти монолинейным ' +
      'штрихом. В горизонтальном наборе во всю ширину экрана это его родная стихия: ' +
      'круглые О и С держат разрядку в 200 px, слово не рассыпается на значки. ' +
      'Кириллица спокойная — Д без выносных лап, Ж собранная, З без завитка. ' +
      'На Тильде такого нет: гарнитура молодая и читается как современная айдентика, ' +
      'а не как ресторанная вывеска. Onest под текст — родная кириллица, девять весов, ' +
      'высокий очковый размер, держит 11 px с трекингом.',
  },
  {
    id: 'podkova',
    title: 'Podkova × Manrope',
    display: 'podkova',
    displayWeight: 700,
    text: 'manrope',
    foundry: 'Cyreal × Михаил Шаранда',
    designer: 'Илья Юдин · Михаил Шаранда',
    link: 'https://fonts.google.com/specimen/Podkova',
    price: 'Бесплатно, SIL OFL 1.1',
    licence: 'SIL Open Font License 1.1',
    cyr: 'Рисовалась в Cyreal под кириллицу, латиница вторична. По cmap 68 из 68.',
    rationale:
      'Брусковая антиква с короткими плотными засечками и низким контрастом. ' +
      'Самая маленькая высота прописных из трёх (0.59 em) — значит на той же высоте ' +
      'литеры слово занимает больше вертикали и обрезка нижней кромкой читается ' +
      'мягче. Характер тёплый и тяжёлый, ближе к дереву и камню, чем к стеклу: ' +
      'для термального комплекса это попадание в материал. Риск: брусковые засечки ' +
      'на очень широком наборе начинают спорить с воздухом между литерами — ' +
      'смотреть надо именно на вордмарке, что витрина и позволяет.',
  },
]

/** В бою — Golos Text, по прямому указанию арт-директора. */
export const DEFAULT_PAIR = PAIRS[0]

File: scripts/fonts.py
#!/usr/bin/env python3
"""
Готовит шрифты: сабсет под кириллицу + woff2 + метрики.

Метрики нужны не для красоты. Вордмарк TERMA обрезается нижней кромкой
вьюпорта ровно на треть высоты литеры — чтобы посадить его точно, надо
знать capHeight и положение базовой линии конкретной гарнитуры, а не
подгонять magic-number на глаз.
"""
import json
import os
import subprocess
import sys

from fontTools.ttLib import TTFont

HERE = os.path.dirname(os.path.abspath(__file__))
SRC = os.environ.get("TERMA_TTF", "/workspace/fontlab/ttf")
# Лицензионные исходники живут отдельно и в репозиторий не попадают:
# публичный репозиторий — не место для полного TTF купленной гарнитуры.
# В выдачу уходит только сабсет woff2, он и закоммичен.
LICENSED = os.path.join(HERE, "..", "assets-src", "fonts")
OUT = os.path.join(HERE, "..", "public", "fonts")
DATA = os.path.join(HERE, "..", "src", "lib", "fonts.generated.json")

# Кириллица целиком + латиница (только для имени бренда и техстрок) +
# цифры, пунктуация, ₽, °, №, —, тире, кавычки-ёлочки.
UNICODES = (
    "U+0020-007E,U+00A0,U+00B0,U+00AB,U+00BB,U+2010-2015,U+2018-201F,"
    "U+2026,U+2116,U+20BD,U+0400-045F,U+0490-0491,U+2212,U+00D7"
)

# Forum, Prata и Tenor Sans выведены из проекта: арт-директор забраковал их
# как самые заезженные «премиальные» гарнитуры русского веба.
# До выбора новой пары Golos Text работает и текстом, и акциденцией —
# в тяжёлом весе он держит крупный кегль честнее любой антиквы с Тильды.
FAMILIES = [
    # id, файл, семейство CSS, роль, вес/оси
    ("kudryashev", "KudryashevDisplay.ttf", "Kudryashev Display", "display", None),
    ("golos",     "GolosText_wght.ttf",    "Golos Text", "text",    (400, 900)),
    ("unbounded", "Unbounded_wght.ttf",    "Unbounded",  "display", (200, 900)),
    ("geologica", "Geologica_CRSV_SHRP_slnt_wght.ttf", "Geologica", "display", (100, 900)),
    ("onest",     "Onest_wght.ttf",        "Onest",      "text",    (100, 900)),
    ("manrope",   "Manrope_wght.ttf",      "Manrope",    "text",    (200, 800)),
    ("podkova",   "Podkova_wght.ttf",      "Podkova",    "display", (400, 800)),
]


def metrics(path):
    f = TTFont(path, lazy=True)
    upm = f["head"].unitsPerEm
    os2 = f["OS/2"]
    hhea = f["hhea"]
    cap = getattr(os2, "sCapHeight", None) or int(upm * 0.7)
    asc, desc = hhea.ascender, hhea.descender
    # Ширины литер T E R M A при весе по умолчанию — для проверки, что
    # слово вообще собирается в очень широком наборе без наездов.
    hmtx = f["hmtx"]
    cmap = f.getBestCmap()
    adv = {}
    for ch in "TERMA":
        gn = cmap.get(ord(ch))
        if gn:
            adv[ch] = round(hmtx[gn][0] / upm, 4)
    f.close()
    return dict(
        upm=upm,
        capR=round(cap / upm, 4),
        ascR=round(asc / upm, 4),
        descR=round(desc / upm, 4),
        adv=adv,
    )


def main():
    os.makedirs(OUT, exist_ok=True)
    out = {}
    for fid, fname, family, role, axes in FAMILIES:
        src = os.path.join(SRC, fname)
        if not os.path.exists(src):
            src = os.path.join(LICENSED, fname)
        if not os.path.exists(src):
            # Лицензионного исходника нет — оставляем уже собранный сабсет.
            dst = os.path.join(OUT, f"{fid}.woff2")
            if os.path.exists(dst) and os.path.exists(DATA):
                prev = json.load(open(DATA, encoding="utf-8"))
                if fid in prev:
                    out[fid] = prev[fid]
                    print(f"{fid:10} исходника нет — оставлен готовый сабсет")
                    continue
            sys.exit(f"нет исходника: {fname}")
        dst = os.path.join(OUT, f"{fid}.woff2")
        cmd = [
            sys.executable, "-m", "fontTools.subset", src,
            f"--unicodes={UNICODES}",
            "--layout-features=kern,liga,calt,locl,onum,tnum,case",
            "--flavor=woff2",
            "--desubroutinize",
            f"--output-file={dst}",
        ]
        subprocess.run(cmd, check=True, capture_output=True)
        m = metrics(src)
        out[fid] = dict(
            id=fid, family=family, role=role, file=f"{fid}.woff2",
            variable=bool(axes), wght=list(axes) if axes else [400],
            bytes=os.path.getsize(dst), **m,
        )
        print(f"{fid:10} {family:12} {out[fid]['bytes']:>7} B  cap={m['capR']}")

    with open(DATA, "w", encoding="utf-8") as fh:
        json.dump(out, fh, ensure_ascii=False, indent=1)
    print("метрики →", DATA)


if __name__ == "__main__":
    main()
